src/client/Client.h

```diff
@@ -1065,6 +1065,7 @@ class Client : public Dispatcher, public md_config_obs_t {
    */
   void _finish_init();
 
+  ceph::spinlock ll_snap_ref_lock;
```

Comment:
Userland spinlocks are somewhat unpredictable as you can get scheduled off the CPU while holding one, in which case you can end up taking a rather nasty performance hit -- other locking threads will still be spinning on the CPUs until the holder is scheduled again and can release it. In userland, the general wisdom is to use mutexes instead of spinlocks unless you have a *very* good reason. I'm not sure this qualifies.

Reply:
> Userland spinlocks are somewhat unpredictable as you can get scheduled off the CPU while holding one, in which case you can end up taking a rather nasty performance hit -- other locking threads will still be spinning on the CPUs until the holder is scheduled again and can release it.

Yeah, sounds reasonable.

> In userland, the general wisdom is to use mutexes instead of spinlocks unless you have a _very_ good reason. I'm not sure this qualifies.

While the `ll_snap_ref_lock` spin lock here will only protect `ll_snap_ref` map and the critical section is very simple and short. The case you mentioned should be okay ?

Reply:
@jtlayton 

I am also thinking whether will it make sense when the `fromdir` and `todir` have no parent/child relationship, can we just use the `Inode::vino` to order the locks, let's always make the smaller one in front the other one when taking the lock ? 

The `Inode::vino` won't change in Inode's life time, then we can get rid of the `s_vfs_rename_mutex` global lock.

Could you foresee any potential issue for this ?

Reply:
Simple and small is good for a spinlock-protected critical, but again, it can be hard to predict how they will perform. I generally recommend avoiding them in userland code unless you *really* know what you're doing.

Basing the locking on the Inode::vino does sound reasonable. I'll think about it a bit more though.

Reply:
> Simple and small is good for a spinlock-protected critical, but again, it can be hard to predict how they will perform. I generally recommend avoiding them in userland code unless you _really_ know what you're doing.
> 

Yeah, you are right. 
I think it is hard to predict what the `map/list/vector, etc` internally will do, maybe allocating new memories which could cause sleeping.

> Basing the locking on the Inode::vino does sound reasonable. I'll think about it a bit more though.

Sure, thanks.
Locally the rename lock patch is ready and can reduce many lock thrashing :-)